File: ruby/find_first_duplicate.rb
require 'set'

def find_first_duplicate(arr)
  unique_values = Set.new

  arr.each do |value|
    return value if unique_values.include?(value)

    unique_values.add(value)
  end

  -1
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  const uniqueValues = new Set()
  for (const value of arr) {
    if (uniqueValues.has(value)) {
      return value
    }
    uniqueValues.add(value)
  }
  return -1
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;